spectral_cube/spectral_cube.py

```diff
@@ -2580,69 +2585,123 @@ def spectral_smooth_median(self, ksize, update_function=None, **kwargs):
+        # 'spectra' is a generator
+        # the boolean check will skip the function for bad spectra
         # TODO: should spatial good/bad be cached?
-        cubelist = ((self.filled_data[:,jj,ii],
-                     self.mask.include(view=(slice(None), jj, ii)))
-                    for jj in range(self.shape[1])
-                    for ii in range(self.shape[2]))
+        spectra = ((self.filled_data[:,jj,ii],
+                    self.mask.include(view=(slice(None), jj, ii)),
+                    ii, jj,
+                   )
+                   for jj in range(shape[1])
+                   for ii in range(shape[2]))
 
         if update_function is None:
             pb = ProgressBar(shape[1] * shape[2])
             update_function = pb.update
 
-        def _gsmooth_spectrum(args):
+        if use_memmap:
+            ntf = tempfile.NamedTemporaryFile()
+            outcube = np.memmap(ntf, mode='w+', shape=shape, dtype=np.float)
+        else:
+            if self._is_huge and not self.allow_huge_operations:
+                raise ValueError("Applying a function without ``use_memmap`` "
+                                 "requires loading the whole array into "
+                                 "memory *twice*, which can overload the "
+                                 "machine's memory for large cubes.  Either "
+                                 "set ``use_memmap=True`` or set "
+                                 "``cube.allow_huge_operations=True`` to "
+                                 "override this restriction.")
+            outcube = np.empty(shape=shape, dtype=np.float)
+
+        def _apply_function(arguments, outcube=outcube, kwargs=kwargs):
             """
             Helper function to smooth a spectrum
             """
-            (spec, includemask),kwargs = args
+            (spec, includemask, ii, jj) = arguments
             update_function()
 
             if any(includemask):
-                return ndimage.filters.median_filter(spec, size=ksize)
+                outcube[:,jj,ii] = function(spec, includemask, **kwargs)
             else:
-                return spec
+                outcube[:,jj,ii] = spec
 
-        # could be numcores, except _gsmooth_spectrum is unpicklable
-        with cube_utils._map_context(1) as map:
-            smoothcube_ = np.array([x for x in
-                                    map(_gsmooth_spectrum, zip(cubelist,
-                                                               itertools.cycle([kwargs]),
-                                                              )
-                                       )
-                                   ]
-                                  )
-
-        # empirical: need to swapaxes to get shape right
-        # cube = np.arange(6*5*4).reshape([4,5,6]).swapaxes(0,2)
-        # cubelist.T.reshape(cube.shape) == cube
-        smoothcube = smoothcube_.T.reshape(shape)
+        # TEST
+        from joblib import Parallel, delayed
```

Comment:
Will the map context manager eventually be used here?

Reply:
Actually, no, my plan is to factor out the map context manager entirely; it was a failure.  multiprocess pool mapping didn't work for any of the intended use cases.